File: ch07/CS/Data/Data/DataForm.cs
﻿using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace Data
{
    public partial class DataForm : Form
    {
        public DataForm()
        {
            InitializeComponent();
        }

        private ISource currentSource;
        private ObjectSource objectSource;
        private DataSetSource dataSetSource;
        private BindingSource categoriesBindingSource = new BindingSource();
        private BindingSource productsBindingSource = new BindingSource();

        private void DataForm_Load(object sender, EventArgs e)
        {
            try
            {
                sourceToolStripComboBox.ComboBox.SelectedIndex = 0;
                SetSource();
                BindCategories();

                categoriesBindingSource.DataSource = currentSource.GetCategories();

                categoryToolStripComboBox.ComboBox.DisplayMember = "CategoryName";
                categoryToolStripComboBox.ComboBox.ValueMember = "CategoryID";
                categoryToolStripComboBox.ComboBox.DataSource = categoriesBindingSource;

                productsListBox.DisplayMember = "ProductName";
                productsListBox.ValueMember = "ProductID";
                productsListBox.DataSource = productsBindingSource;
                productDataGridView.DataSource = productsBindingSource;

                nameTextBox.DataBindings.Add("Text", productsBindingSource, "ProductName");
                quantityPerUnitTextBox.DataBindings.Add("Text", productsBindingSource, "QuantityPerUnit");
                priceTextBox.DataBindings.Add("Text", productsBindingSource, "UnitPrice");
                stockTextBox.DataBindings.Add("Text", productsBindingSource, "UnitsInStock");
                orderTextBox.DataBindings.Add("Text", productsBindingSource, "UnitsOnOrder");
                discontinuedCheckBox.DataBindings.Add("Checked", productsBindingSource, "Discontinued");
            }
            catch (Exception ex)
            {
                MessageBox.Show(ex.Message, this.Text);
                throw;
            }
        }

        private void categoryToolStripComboBox_SelectedIndexChanged(object sender, EventArgs e)
        {
            BindProducts();
        }

        private void BindProducts()
        {
            var catId = Convert.ToInt32(categoryToolStripComboBox.ComboBox.SelectedValue);
            productsBindingSource.DataSource = currentSource.GetProducts(catId);
        }

        private void BindCategories()
        {
            categoriesBindingSource.DataSource = currentSource.GetCategories();
        }

        private void SetSource()
        {
            switch (sourceToolStripComboBox.ComboBox.SelectedIndex)
            {
                case 0:
                    if (objectSource == null)
                    {
                        objectSource = new ObjectSource();
                    }
                    currentSource = objectSource;
                    break;
                case 1:
                    if (dataSetSource == null)
                    {
                        dataSetSource = new DataSetSource();
                    }
                    currentSource = dataSetSource;
                    break;
                default:
                    break;
            }
        }

        private void AddToolStripButton_Click(object sender, EventArgs e)
        {
            var id = Convert.ToInt32(categoryToolStripComboBox.ComboBox.SelectedValue);
            var name = categoryToolStripComboBox.ComboBox.Text;
            var category = new Category(id, name);
            var form = new AddProductForm(category);
            var result = form.ShowDialog();
            if (result == DialogResult.OK)
            {
                currentSource.AddProduct(productsBindingSource, form.Product);
            }
        }

        private void DeleteToolStripButton_Click(object sender, EventArgs e)
        {
            if (productsListBox.SelectedValue != null)
            {
                currentSource.DeleteProduct(productsBindingSource, (int)productsListBox.SelectedValue);
            }
        }

        private void BackToolStripButton_Click(object sender, EventArgs e)
        {
            productsBindingSource.MovePrevious();
        }

        private void ForwardToolStripButton_Click(object sender, EventArgs e)
        {
            productsBindingSource.MoveNext();
        }

        private void sourceToolStripComboBox_SelectedIndexChanged(object sender, EventArgs e)
        {
            SetSource();
            BindCategories();
            BindProducts();
        }

        private void SaveToolStripButton_Click(object sender, EventArgs e)
        {
            currentSource.Save();
        }
    }
}

File: ch07/VB/Data/Data/DataForm.vb
Public Class DataForm
    Private _categoriesBindingSource As New BindingSource()
    Private _productsBindingSource As New BindingSource()

    Private _currentSource As ISource
    Private _objectSource As ObjectSource
    Private _dataSetSource As DataSetSource

    Private Sub DataForm_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        SourceToolStripComboBox.SelectedIndex = 0
        SetSource()
        BindCategories()

        _categoriesBindingSource.DataSource = _currentSource.GetCategories()

        CategoryToolStripComboBox.ComboBox.DisplayMember = "CategoryName"
        CategoryToolStripComboBox.ComboBox.ValueMember = "CategoryID"
        CategoryToolStripComboBox.ComboBox.DataSource = _categoriesBindingSource

        ProductsListBox.DataSource = _productsBindingSource
        ProductsListBox.DisplayMember = "ProductName"
        ProductsDataGridView.DataSource = _productsBindingSource

        NameTextBox.DataBindings.Add("Text", _productsBindingSource, "ProductName")
        QuantityPerUnitTextBox.DataBindings.Add("Text", _productsBindingSource, "QuantityPerUnit")
        PriceTextBox.DataBindings.Add("Text", _productsBindingSource, "UnitPrice")
        StockTextBox.DataBindings.Add("Text", _productsBindingSource, "UnitsInStock")
        UnitsOnOrderTextBox.DataBindings.Add("Text", _productsBindingSource, "UnitsOnOrder")
        DiscontinuedCheckBox.DataBindings.Add("Checked", _productsBindingSource, "Discontinued")
    End Sub

    Private Sub CategoryToolStripComboBox_SelectedIndexChanged(sender As Object, e As EventArgs) Handles CategoryToolStripComboBox.SelectedIndexChanged
        BindProducts()
    End Sub

    Private Sub AddToolStripButton_Click(sender As Object, e As EventArgs) Handles AddToolStripButton.Click
        Dim category = CType(CategoryToolStripComboBox.SelectedItem, Category)
        Dim form As New AddProductForm(category)
        Dim result = form.ShowDialog()
        If result = DialogResult.OK Then
            _productsBindingSource.Add(form.Product)
        End If

    End Sub

    Private Sub DeleteToolStripButton_Click(sender As Object, e As EventArgs) Handles DeleteToolStripButton.Click
        Dim product = CType(ProductsListBox.SelectedItem, Product)
        _productsBindingSource.Remove(product)
    End Sub

    Private Sub BackToolStripButton_Click(sender As Object, e As EventArgs) Handles BackToolStripButton.Click
        _productsBindingSource.MovePrevious()
    End Sub

    Private Sub ForwardToolStripButton_Click(sender As Object, e As EventArgs) Handles ForwardToolStripButton.Click
        _productsBindingSource.MoveNext()
    End Sub

    Private Sub SourceToolStripComboBox_SelectedIndexChanged(sender As Object, e As EventArgs) Handles SourceToolStripComboBox.SelectedIndexChanged
        SetSource()
        BindCategories()
        BindProducts()
    End Sub

    Private Sub BindCategories()
        _categoriesBindingSource.DataSource = _currentSource.GetCategories()
    End Sub

    Private Sub BindProducts()
        Dim catId = CategoryToolStripComboBox.ComboBox.SelectedValue
        _productsBindingSource.DataSource = _currentSource.GetProducts(catId)
    End Sub

    Private Sub SetSource()
        Select Case SourceToolStripComboBox.SelectedIndex
            Case 0
                If _objectSource Is Nothing Then
                    _objectSource = New ObjectSource()
                End If
                _currentSource = _objectSource
            Case 1
                If _dataSetSource Is Nothing Then
                    _dataSetSource = New DataSetSource()
                End If
                _currentSource = _dataSetSource
        End Select
    End Sub
End Class
